Optimize and multithread remove diagonal before measure pass

mtreinish opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add a release note file using reno if the change needs to be
  documented in the release notes.
⚠️ If your pull request fixes an open issue, please link to the issue.

- [ ] I have added the tests to cover my changes.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the CONTRIBUTING document.
-->

### Summary

This commit reworks the `RemoveDiagonalGatesBeforeMeasure` transpiler implementation to be more efficient by removing string comparisons to identify measurements and simplify the code while doing this. This changes the measurement identification to be a u8 comparisons vs a string comparison looking for "measure". At the same time the new form was conducive for multithreading using rayon. The pass has two stages one to analyze all the nodes in the dag looking for measurements and checking if their direct predecessors are diagonal gates or not. Then one stage to remove all the identified diagonal gates. The first stage has no data dependencies between the checks and this analysis can be done in parallel. This commit adds a rayon iterator to do the analysis in parallel.

### Details and comments

TODO:

- [x] Scale test and find crossover point to switch parallel and serial execution

mtreinish committed:
Optimize and multithread remove diagonal before measure pass

This commit reworks the RemoveDiagonalGatesBeforeMeasure transpiler
implementation to be more efficient by removing string comparisons to
identifiy measurements and simplify the code while doing this. This
changes the measurement identification to be a u8 comparisons vs a
string comparison looking for "measure". At the same time the new form
was conducive for multithreading using rayon. The pass has two stages
one to analyze all the nodes in the dag looking for measurements and
checking if their direct predecessors are diagonal gates or not. Then
one stage to remove all the identified diagonal gates. The first stage
has no data dependencies between the checks and this analysis can be
done in parallel. This commit adds a rayon iterator to do the analysis
in parallel.

mtreinish commented:
I've done the scale testing with:
```python
import csv
import statistics
import time

import numpy as np

from qiskit.circuit import QuantumCircuit
from qiskit.converters import circuit_to_dag
from qiskit.circuit.quantumcircuitdata import CircuitInstruction
from qiskit.circuit.library import XGate, RZGate, RGate
from qiskit.transpiler.passes import RemoveDiagonalGatesBeforeMeasure

opt_pass = RemoveDiagonalGatesBeforeMeasure()

with open("times_serial_diag.csv", "w") as fd:
    writer = csv.writer(fd)
    writer.writerow(["circuit", "size", "runtime"])
    for size in np.logspace(1, 7, dtype=int):
        size = int(size)
        c_no_measure = QuantumCircuit(1, 1)
        qubits = [c_no_measure.qubits[0]]
        for _ in range(size):
            c_no_measure._append(CircuitInstruction.from_standard(XGate._standard_gate, qubits, []))
        c_no_diag = QuantumCircuit(1, 1)
        qubits = [c_no_diag.qubits[0]]
        for _ in range(size):
            c_no_diag._append(CircuitInstruction.from_standard(XGate._standard_gate, qubits, []))
            c_no_diag.measure(0, 0)
        c_half_diag = QuantumCircuit(1)
        qubits = [c_half_diag.qubits[0]]
        for _ in range(int(size / 2)):
            c_half_diag._append(CircuitInstruction.from_standard(XGate._standard_gate, qubits, []))
            c_no_diag.measure(0, 0)
        for _ in range(int(size / 2)):
            c_half_diag._append(CircuitInstruction.from_standard(RZGate._standard_gate, qubits, [0.]))
            c_no_diag.measure(0, 0)
        c_full_diag = QuantumCircuit(1)
        qubits = [c_full_diag.qubits[0]]
        for _ in range(size):
            c_full_diag._append(CircuitInstruction.from_standard(RZGate._standard_gate, qubits, [0.]))
            c_no_diag.measure(0, 0)

        d_no_diag = circuit_to_dag(c_no_diag)
        d_half_diag = circuit_to_dag(c_half_diag)
        d_full_diag = circuit_to_dag(c_full_diag)
        d_no_measure = circuit_to_dag(c_no_measure)
        times = []
        for _ in range(10):
            start = time.perf_counter()
            opt_pass.run(d_no_diag)
            stop = time.perf_counter()
            times.append(stop - start)
        writer.writerow(["None", size, statistics.geometric_mean(times)])
        print(f"No diag with measure size {size}: {statistics.geometric_mean(times)}")
        times = []
        for _ in range(10):
            start = time.perf_counter()
            opt_pass.run(d_half_diag)
            stop = time.perf_counter()
            times.append(stop - start)
        writer.writerow(["Half", size, statistics.geometric_mean(times)])
        print(f"Half size diag {size}: {statistics.geometric_mean(times)}")
        times = []
        for _ in range(10):
            start = time.perf_counter()
            opt_pass.run(d_full_diag)
            stop = time.perf_counter()
            times.append(stop - start)
        writer.writerow(["Full", size, statistics.geometric_mean(times)])
        print(f"Full size diag {size}: {statistics.geometric_mean(times)}")
        times = []
        for _ in range(10):
            start = time.perf_counter()
            opt_pass.run(d_no_measure)
            stop = time.perf_counter()
            times.append(stop - start)
        writer.writerow(["NoMeasure", size, statistics.geometric_mean(times)])
        print(f"No measure {size}: {statistics.geometric_mean(times)}")

        fd.flush()
```

To compare the pass speed with different circuit configurations. I'm not sure what to do with regards as to when we should run this pass multithreaded vs serially. The results are pretty mixed and it only shows a clear advantage in the case when there are no diagonals before any measurements or at >=10 million measurements in the circuit. This might just be an issue with my testing methodology here I was looking for the pathological case to find the worst case scaling and maybe I should benchmark more realistic workloads.

<img width="5760" height="4320" alt="times_remove_diag_Full" src="https://github.com/user-attachments/assets/9cab47fe-b38f-4c86-8906-e4278ac36dbc" />
<img width="5760" height="4320" alt="times_remove_diag_Half" src="https://github.com/user-attachments/assets/1b256c11-b32f-4865-8bf4-e19600668956" />
<img width="5760" height="4320" alt="times_remove_diag_NoMeasure" src="https://github.com/user-attachments/assets/c3e71ce6-8e2f-49ac-9b73-c00f1decef76" />
<img width="5760" height="4320" alt="times_remove_diag_None" src="https://github.com/user-attachments/assets/06b290ba-729b-4d17-a0a3-ef32ae9f4e1a" />

mtreinish commented on crates/transpiler/src/passes/remove_diagonal_gates_before_measure.rs:
```diff
@@ -46,44 +51,61 @@ pub fn run_remove_diagonal_before_measure(dag: &mut DAGCircuit) -> PyResult<()>
+                return None;
+            };
+            if let Some(gate) = pred_inst.standard_gate() {
+                if DIAGONAL_1Q_GATES.contains(&gate) {
```
I'm thinking using `matches!(gate,  ...)` here instead of `contains()` over a static array will end up with a more efficient check at the cost of more verbose code. I'll benchmark pivot the logic here to use that instead if it's measurable.

mtreinish replied:
Done in: https://github.com/Qiskit/qiskit/pull/14723/commits/6c9d71c7326087a2a6c94a4a61164328fb21e060

mtreinish committed:
Use match instead of contains for detection

mtreinish committed:
Skip iteration if there are no measurements

mtreinish committed:
Set threshold for parallelism based on scaling tests

mtreinish commented:
@jakelishman pointed out there was a copy paste error in the above script I used for scale testing. I was only adding the measurements to a single circuit and not all 3 separate circuits as was intended. After fixing that the plots make a lot more sense:

<img width="5760" height="4320" alt="times_remove_diag_Full" src="https://github.com/user-attachments/assets/ebb48776-c5d3-4037-9eaa-49fce0fb3d06" />
<img width="5760" height="4320" alt="times_remove_diag_Half" src="https://github.com/user-attachments/assets/01b18e74-a829-4652-8298-8139d5820cae" />
<img width="5760" height="4320" alt="times_remove_diag_None" src="https://github.com/user-attachments/assets/92cc62a9-baa9-43d0-b705-a9395b069fe3" />
<img width="5760" height="4320" alt="times_remove_diag_NoMeasure" src="https://github.com/user-attachments/assets/119de6fe-36e6-49e1-9de6-13cd3caddb49" />

Based on these results I set the threshold for parallelism at 50,000 operations in the circuit. The no measurement case looks completely different because I realized we weren't short circuiting if the circuit didn't have any measurements which I added in: https://github.com/Qiskit/qiskit/pull/14723/commits/7ad9298afb86035430b71ae62eef3c7a090a31e1 I'm not sure why the parallel test case was consistently slower there it might have just been system noise because it's in the < 1 μs scale.

mtreinish committed:
Remove unused result and unnecessary node lookup on removal

mtreinish committed:
Restore check for node in graph

This commit restores the check for the node being removed's presence in
the DAG. THis was actually needed because in the case of 2 or 3q
diagonal gates a gate can end up in the list > 1 time. This will result
in the node trying to be removed more than once, the options are to
either deduplicate the list (by using a hashset or indexset instead of a
Vec) or check before inserting in the dag. A vec will be generally
faster for iteration than a set type and the lookup for the dag is the
equivalent basically doing something like `dag.nodes[index].is_some()`
where `dag.nodes` is a vec which will likely be faster than hashing the
node on collection.

mtreinish committed:
Filter directives as they're not measurements

jakelishman commented on crates/transpiler/src/passes/remove_diagonal_gates_before_measure.rs:
```diff
@@ -24,74 +29,87 @@ use qiskit_circuit::operations::StandardGate;
+                return None;
+            };
+            if let Some(gate) = pred_inst.standard_gate() {
+                match gate {
+                    StandardGate::RZ
+                    | StandardGate::Z
+                    | StandardGate::T
+                    | StandardGate::S
+                    | StandardGate::Tdg
+                    | StandardGate::Sdg
+                    | StandardGate::U1
+                    | StandardGate::Phase => return Some(predecessor),
+                    StandardGate::CZ
+                    | StandardGate::CRZ
+                    | StandardGate::CU1
+                    | StandardGate::RZZ
+                    | StandardGate::CPhase
+                    | StandardGate::CS
+                    | StandardGate::CSdg
+                    | StandardGate::CCZ => {
+                        let mut successors = dag.quantum_successors(predecessor);
+                        if successors.all(|s| {
+                            let node_s = &dag.dag()[s];
+                            if let NodeType::Operation(inst_s) = node_s {
+                                matches!(
+                                    inst_s.op.view(),
+                                    OperationRef::StandardInstruction(StandardInstruction::Measure)
+                                )
+                            } else {
+                                false
                             }
+                        }) {
+                            return Some(predecessor);
                         }
```
I don't know if LLVM will codegen this, but since `StandardGate` is a `u8` and we're only matching on two lots of eight of them, this whole check is probably doable in a couple of SSE2 instructions (load a hardcoded vector of all relevant `u8`s, splat the current `StandardGate` into a vector, compare for equality, branch on zeroness of output).

mtreinish replied:
That's why I switched over to use a match because I had hoped that LLVM would be smart enough to recognize this and use SIMD instructions. I can check the generated assembly and use pulp or something if it doesn't.

jakelishman commented on crates/transpiler/src/passes/remove_diagonal_gates_before_measure.rs:
```diff
@@ -24,74 +29,87 @@ use qiskit_circuit::operations::StandardGate;
-        if dag.dag().node_weight(node_to_remove).is_some() {
-            dag.remove_op_node(node_to_remove);
+    let nodes_to_remove: Vec<NodeIndex> = if run_in_parallel && dag.num_ops() >= 50_000 {
+        let node_indices = dag.dag().node_indices().collect::<Vec<_>>();
+        node_indices
+            .into_par_iter()
```
I feel like the trick here is really going to be that we're adding an extra serialised collection of all the node indices, which is going to be a pretty large offset, especially for the vast majority of circuits, which have really really few measurements, and just in terminal positions.  I wonder if you'd see any difference if you move the "is this a measurement?" filter part into the collection of line 91.

Going further: if this is something we're going to need more of in the future, we really ought to look into providing Rayon-enabled `topological_op_nodes` and `op_nodes` methods on `DAGCircuit` from Rust space.  It's plausible we'd need to PR to `petgraph` to get some of the underlying infrastructure in place for things like `node_indices` on `StableGraph` (since for `Graph`, we could easily make do without Rayon support in `petgraph` - we can just generate the indices ourself from a range).

mtreinish replied:
Yeah I reached the same conclusion that realistically we'll need to do add it to petgraph to run without an allocation (I did try `ParBridge` but the performance was abysmal). I actually have a draft in progress of contributing a `par_node_indices` for `Graph` and `StableGraph` locally that's not quite compiling yet (some trait bound is missing).

I debated doing the measurement filtering outside the parallelism but figured it was better to do it inside. Since we'd be doing a lot of checking serially vs one much larger allocation. I can benchmark it to make sure though.

jakelishman commented on crates/transpiler/src/passes/remove_diagonal_gates_before_measure.rs:
```diff
@@ -24,74 +29,87 @@ use qiskit_circuit::operations::StandardGate;
+            .collect()
+    };
+
+    for node in nodes_to_remove {
+        if dag.dag().node_weight(node).is_some() {
+            dag.remove_op_node(node);
         }
     }
```
Is the `is_some()` check here to avoid 2q gates getting removed more than once?

mtreinish replied:
Yeah see: https://github.com/Qiskit/qiskit/pull/14723/commits/c041937cd9a5eb919f64179c99293176759b84c9 where I added it back after removing it.

jakelishman reviewed (commented):
I think your testing script at the moment either makes 50% of the gates measurements, or none of them.  I think more typical circuits are going to have many many operations, followed by exactly one measurement on each qubit.  Can you see if your speed-ups still hold in those sorts of circuit?

mtreinish commented:
> I think your testing script at the moment either makes 50% of the gates measurements, or none of them. I think more typical circuits are going to have many many operations, followed by exactly one measurement on each qubit. Can you see if your speed-ups still hold in those sorts of circuit?

I did it this way primarily to have worst case numbers for the pass since I thought that would be the most interesting. But you're right it's not a realistic measure. I'll try it with a 200 qubit circuit of increasing depth that only has 200 final measurements and see what the scaling is like there.

mtreinish committed:
Avoid vec collection of indices

mtreinish committed:
Merge remote-tracking branch 'origin/main' into diagonal-parallel-remove

mtreinish committed:
Fix rebase issues

jakelishman committed:
Refactor to flatten nesting (#33)

jakelishman committed:
Fix incorrect index in successors

mtreinish commented:
I've updated my test script per the ealier review comment: https://github.com/Qiskit/qiskit/pull/14723#pullrequestreview-3015835013 to use terminal measurements on all qubits (where the number of qubits is $\sqrt{n_{gates}}$)

```python
import csv
import statistics
import time

import numpy as np

from qiskit.circuit import QuantumCircuit
from qiskit.converters import circuit_to_dag
from qiskit.circuit.quantumcircuitdata import CircuitInstruction
from qiskit.circuit.library import XGate, RZGate, RGate
from qiskit.transpiler.passes import RemoveDiagonalGatesBeforeMeasure

opt_pass = RemoveDiagonalGatesBeforeMeasure()

with open("times_serial_bridge_diag.csv", "w") as fd:
    writer = csv.writer(fd)
    writer.writerow(["circuit", "size", "runtime"])
    for size in np.logspace(1, 3.2, dtype=int):
        size = int(size)
        c_no_measure = QuantumCircuit(size, size)
        for i in range(size):
            for _ in range(size):
                c_no_measure._append(CircuitInstruction.from_standard(XGate._standard_gate, [c_no_measure.qubits[i]], []))

        c_no_diag = QuantumCircuit(size, size)
        for i in range(size):
            for _ in range(size):
                c_no_diag._append(CircuitInstruction.from_standard(XGate._standard_gate, [c_no_diag.qubits[i]], []))
            c_no_diag.measure(i, i)
        c_half_diag = QuantumCircuit(size, size)
        for i in range(size):
            qubits = [c_half_diag.qubits[i]]
            for j in range(size):
                if i % 2:
                    if j < size - 2:
                        c_half_diag._append(CircuitInstruction.from_standard(XGate._standard_gate, qubits, []))
                    else:
                        c_half_diag._append(CircuitInstruction.from_standard(RZGate._standard_gate, qubits, [0.]))
                else:
                    c_half_diag._append(CircuitInstruction.from_standard(XGate._standard_gate, qubits, []))
        for i in range(size):
            c_half_diag.measure(i, i)
        c_full_diag = QuantumCircuit(size, size)
        for i in range(size):
            qubits = [c_full_diag.qubits[i]]
            for _ in range(size):
                c_full_diag._append(CircuitInstruction.from_standard(RZGate._standard_gate, qubits, [0.]))
            c_full_diag.measure(i, i)

        d_no_diag = circuit_to_dag(c_no_diag)
        d_half_diag = circuit_to_dag(c_half_diag)
        d_full_diag = circuit_to_dag(c_full_diag)
        d_no_measure = circuit_to_dag(c_no_measure)
        times = []
        for _ in range(25):
            start = time.perf_counter()
            opt_pass.run(d_no_diag)
            stop = time.perf_counter()
            times.append(stop - start)
        writer.writerow(["None", size, statistics.geometric_mean(times)])
        print(f"No diag with measure size {size}: {statistics.geometric_mean(times)}")
        times = []
        for _ in range(25):
            start = time.perf_counter()
            opt_pass.run(d_half_diag)
            stop = time.perf_counter()
            times.append(stop - start)
        writer.writerow(["Half", size, statistics.geometric_mean(times)])
        print(f"Half size diag {size}: {statistics.geometric_mean(times)}")
        times = []
        for _ in range(25):
            start = time.perf_counter()
            opt_pass.run(d_full_diag)
            stop = time.perf_counter()
            times.append(stop - start)
        writer.writerow(["Full", size, statistics.geometric_mean(times)])
        print(f"Full size diag {size}: {statistics.geometric_mean(times)}")
        times = []
        for _ in range(10):
            start = time.perf_counter()
            opt_pass.run(d_no_measure)
            stop = time.perf_counter()
            times.append(stop - start)
        writer.writerow(["NoMeasure", size, statistics.geometric_mean(times)])
        print(f"No measure {size}: {statistics.geometric_mean(times)}")

        fd.flush()
```

---

Looking at the plotted outputs from this I think the 50k gates number is still valid. So I'll keep that value and document it by citing this comment. But if people would like to verify these numbers on their local environment we can still adjust this before merging.

<img width="5760" height="4320" alt="full_diag" src="https://github.com/user-attachments/assets/b417aba7-e34b-4005-9bbf-40d3909f67d8" />
<img width="5760" height="4320" alt="half_diag" src="https://github.com/user-attachments/assets/6d56248f-09dc-4d57-a621-b3f9bb2c6d8b" />
<img width="5760" height="4320" alt="no_diag" src="https://github.com/user-attachments/assets/82379f23-8a34-46b1-b70f-382265d7884d" />

mtreinish committed:
Document source of parallel threshold

mtreinish committed:
Deduplicate multiqubit gates prior to serial removal loop

This commit updates the logic of the process_node closure to only emit
the multiqubit gate for a single measurement avoiding the need to do
another dag node lookup during the removal loop.

Co-authored-by: Jake Lishman <jake.lishman@ibm.com>

ShellyGarion reviewed (commented):
I think this PR deserves a performance release note.
Was the threshold of 50K verified on independent environments and machines?
